serveradmin/servershell/helper/autocomplete.py

```diff
@@ -0,0 +1,98 @@
+from django.db import connection, ProgrammingError
+
+from serveradmin.serverdb.models import Attribute, ServerAttribute, Server
+from django.core.exceptions import ObjectDoesNotExist
+
+
+def attribute_startswith(search_string, limit=20):
+    """Query attributes starting search_string
+
+    :param search_string: e.g. al to match allow_from, allow_to etc.
+    :param limit: limit result to n results
+
+    :return:
+    """
+
+    query = Attribute.objects.filter(
+        attribute_id__startswith=search_string).only('attribute_id').order_by(
+        'attribute_id')
+    attributes = [attribute.attribute_id for attribute in query[:limit]]
+
+    for attribute in Attribute.specials.keys():
+        if attribute.startswith(search_string):
+            attributes.append(attribute)
+
+    return sorted(attributes[:limit])
+
+
+def attribute_value_startswith(attribute_id, search_string, limit=20):
+    """Query attribute for values starting with search string
+
+    :param attribute_id: e.g. servertype or primary_ip6
+    :param search_string: e.g. 2a00:
+    :param limit: limit result to n results
+    :return:
+    """
+
+    if attribute_id in Attribute.specials.keys():
+        return _specials_value_startswith(attribute_id, search_string, limit)
+    else:
+        return _value_startswith(attribute_id, search_string, limit)
+
+
+def _specials_value_startswith(attribute_id, search_string, limit=20):
+    """Query attribute specials starting with search_string
+
+    :param attribute_id: e.g. servertype
+    :param search_string: e.g. v
+    :param limit: limit result to n results
+    :return:
+    """
+
+    column_name = Attribute.specials[attribute_id].special.field
+
+    with connection.cursor() as cursor:
+        # This is safe because field_name comes from us
```

Comment:
Even if these fields come from us, why do we not use a parameterized query and prepared statement? :)

Reply:
Because the execute function blindly escapes all occurrences of %s with two single quotes to make sure it is 100% robust against SQL injections. This will break the SQL query as you the parameters in the select and where clause are column names and the limit must be a integer.
`
sql = "SELECT DISTINCT %s FROM server WHERE %s LIKE %s ORDER BY %s ASC LIMIT %s"
c.execute(sql, ['hostname', 'hostname', 'de100%', 'hostname', '20'])
django.db.utils.ProgrammingError: non-integer constant in ORDER BY
LINE 1: ...OM server WHERE 'hostname' LIKE 'de100%' ORDER BY 'hostname'...
`